--- a/win32/Makefile.gcc
+++ b/win32/Makefile.gcc
@@ -40,19 +40,20 @@
 SHARED_MODE=0

 #LOC = -DZLIB_DEBUG -g

-PREFIX =
-CC = $(PREFIX)gcc
+HOST =
+CC = $(HOST)-gcc
 CFLAGS = $(LOC) -O3 -Wall

-AS = $(CC)
+AS = $(HOST)-gcc
 ASFLAGS = $(LOC) -Wall

-LD = $(CC)
+LD = $(HOST)-gcc
 LDFLAGS = $(LOC)

-AR = $(PREFIX)ar
+AR = $(HOST)-ar
 ARFLAGS = rcs

-RC = $(PREFIX)windres
+RC = $(HOST)-windres
 RCFLAGS =

-STRIP = $(PREFIX)strip
+STRIP = $(HOST)-strip

 CP = cp -fp
 # If GNU install is available, replace $(CP) with install.
@@ -60,8 +61,11 @@
 INSTALL = $(CP)
 RM = rm -f

-prefix ?= /usr/local
-exec_prefix = $(prefix)
+PREFIX ?= /usr/local
+exec_prefix = $(PREFIX)
+
+BINARY_PATH = $(PREFIX)/bin
+INCLUDE_PATH = $(PREFIX)/include
+LIBRARY_PATH = $(PREFIX)/lib

 OBJS = adler32.o compress.o crc32.o deflate.o gzclose.o gzlib.o gzread.o \
        gzwrite.o infback.o inffast.o inflate.o inftrees.o trees.o uncompr.o zutil.o