--- a/win32/Makefile.gcc	2026-03-16 22:57:27.014415600 +0800
+++ b/win32/Makefile.gcc	2026-03-16 22:58:43.521983900 +0800
@@ -36,8 +36,8 @@

 #LOC = -DZLIB_DEBUG -g

-PREFIX =
-CC = $(PREFIX)gcc
+HOST =
+CC = $(HOST)-gcc
 CFLAGS = $(LOC) -O3 -Wall

 AS = $(CC)
@@ -46,21 +46,25 @@
 LD = $(CC)
 LDFLAGS = $(LOC)

-AR = $(PREFIX)ar
+AR = $(HOST)-ar
 ARFLAGS = rcs

-RC = $(PREFIX)windres
+RC = $(HOST)-windres
 RCFLAGS = --define GCC_WINDRES

-STRIP = $(PREFIX)strip
+STRIP = $(HOST)-strip

 CP = cp -fp
 # If GNU install is available, replace $(CP) with install.
 INSTALL = $(CP)
 RM = rm -f

-prefix ?= /usr/local
-exec_prefix = $(prefix)
+PREFIX ?= /usr/local
+exec_prefix = $(PREFIX)
+
+BINARY_PATH = $(PREFIX)/bin
+INCLUDE_PATH = $(PREFIX)/include
+LIBRARY_PATH = $(PREFIX)/lib

 OBJS = adler32.o compress.o crc32.o deflate.o gzclose.o gzlib.o gzread.o \
        gzwrite.o infback.o inffast.o inflate.o inftrees.o trees.o uncompr.o zutil.o
